espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/espresso/MultiTouchState.java

```diff
@@ -0,0 +1,162 @@
+        if (actionType == ACTION_DOWN) {
+            touchPhase = DOWN;
+        } else if (actionType == ACTION_UP) {
+            touchPhase = UP;
+        }
+    }
+
+    /**
+     * Get the x coordinates for all inputs in the same order they were entered
+     * @return X coordinates as a list
+     */
+    private List<Long> getXCoords() {
+        List<Long> xCoords = new ArrayList<>();
+
+        for(Map.Entry<String, TouchState> entry:touchStateSet.entrySet()) {
+            xCoords.add(entry.getValue().getX());
+        }
+
+        return xCoords;
+    }
+
+    /**
+     * Get the y coordinates for all inputs in the same order they were entered
+     * @return Y coordinates as a list
+     */
+    private List<Long> getYCoords() {
+        List<Long> yCoords = new ArrayList<>(touchStateSet.size());
+
+        for(Map.Entry<String, TouchState> entry:touchStateSet.entrySet()) {
+            yCoords.add(entry.getValue().getY());
+        }
+
+        return yCoords;
+    }
+
+    public void pointerDown(UiController uiController) throws AppiumException {
+        AndroidMotionEvent androidMotionEvent = AndroidMotionEvent.getTouchMotionEvent(uiController);
+        Long eventTime = SystemClock.uptimeMillis();
+        List<Long> xCoords = getXCoords();
+        List<Long> yCoords = getYCoords();
+        this.downEvent = androidMotionEvent.pointerUpOrDown(
+                xCoords, yCoords,
+                ACTION_DOWN, this.button, TOUCH, this.globalKeyInputState, null, eventTime);
+
+        if (xCoords.size() > 1) {
+            androidMotionEvent.pointerUpOrDown(xCoords, yCoords,
+                    ACTION_POINTER_DOWN, this.button, TOUCH, this.globalKeyInputState, downEvent, eventTime);
+        }
+
+    }
+
+    public void pointerUp(UiController uiController) throws AppiumException {
+        AndroidMotionEvent androidMotionEvent = AndroidMotionEvent.getTouchMotionEvent(uiController);
+        Long eventTime = SystemClock.uptimeMillis();
+        List<Long> xCoords = getXCoords();
+        List<Long> yCoords = getYCoords();
+        if (xCoords.size() > 1) {
+            androidMotionEvent.pointerUpOrDown(
+                    xCoords, yCoords,
+                    ACTION_POINTER_UP, this.button, TOUCH, this.globalKeyInputState, downEvent, eventTime);
+        }
+
+        androidMotionEvent.pointerUpOrDown(
+                xCoords, yCoords,
+                ACTION_UP, this.button, TOUCH, this.globalKeyInputState, downEvent, eventTime);
+
+        this.downEvent = null;
+    }
+
+    public void pointerCancel(UiController uiController) throws AppiumException {
+        List<Long> xCoords = getXCoords();
+        List<Long> yCoords = getYCoords();
+        AndroidMotionEvent.getTouchMotionEvent(uiController).pointerCancel(xCoords, yCoords);
+        this.downEvent = null;
+        touchPhase = NONE;
+    }
+
+    public void pointerMove(UiController uiController) throws AppiumException {
+        if (this.isDown()) {
```

Comment:
shall we generate HOVER event if isDown false?

Reply:
Would HOVER be applicable to touch events?

Reply:
no it won't. Only mouse

Reply:
Yeah, this one only applies to touch. I'll look into it for mouse though.